Refactoring config to client — as well as general style and consistency stuff.

hot-leaf-juice opened the pull request:
Started doing this as part of my interview and thought I may as well polish it off and submit a PR. Functionality isn't changed at all, but restructured to a class structure that makes more sense to me and cleaned things up a bit.

topliceanu commented:
Callum, can you make sure the tests pass before I start reviewing, please!

topliceanu commented on pusher/authentication_client.py:
```diff
@@ -0,0 +1,112 @@
+        """
+        key = ensure_text(key, "key")
+        signature = ensure_text(signature, "signature")
+        body = ensure_text(body, "body")
```
`ensure_text` may raise and exception, so `validate_webhook` does as well. This needs to be documented in the Readme! I'm not sure if it's an API change!

hot-leaf-juice replied:
This isn't an API change, I didn't touch the implementation of `ensure_text` or `validate_webhook`, but I can add the documentation to the Readme!

topliceanu commented on pusher/pusher_client.py:
```diff
@@ -0,0 +1,147 @@
+import os
+import re
+import six
+import time
```
I'm not sure what the standard for ordering python imports is. AFAIK it's: - std, - third party, - internal. Wdyt? (I don't feel strongly about this)

hot-leaf-juice replied:
ahh yeah, that's definitely the standard order. This was an oversight on my part – will update!

topliceanu commented on pusher/pusher_client.py:
```diff
@@ -0,0 +1,147 @@
+import time
+
+
+class PusherClient(Client):
```
this is the class that is performing api requests.

topliceanu commented on pusher/signature.py:
```diff
@@ -1,27 +1,36 @@
-        return reduce(lambda x, y: x | y, [ord(x) ^ ord(y) for x, y in zip(a, b)]) == 0
+
+        return reduce(
+            lambda x, y: x | y, [ord(x) ^ ord(y) for x, y in zip(a, b)]) == 0
```
this is not very legible (took me a couple of minutes to understand this one line) but, still, a smart piece of code!

topliceanu commented on setup.py:
```diff
@@ -35,7 +35,7 @@
         'six',
         'requests>=2.3.0',
-        'urllib3',
+        'urllib3==1.17',
```
unfortunate, however it makes the tests pass on travis so it's fine!

topliceanu commented on pusher/authentication_client.py:
```diff
@@ -0,0 +1,112 @@
+        :param channel: name of the channel to authorize subscription to
+        :param socket_id: id of the socket that requires authorization
+        :param custom_data: used on presence channels to provide user info
+        """
```
Can you document what this method returns? It seems to be returning the auth token. If that's the case, it may be that the method's name is not very good.

topliceanu replied:
Ah I see, you split the implementation into NotificationClient, AuthenticationClient and PusherClient. Then, the Pusher object instantiates all three and delegates to them.

hot-leaf-juice commented:
@topliceanu I added the errors to the readme and changed the order of the imports. I didn't write the lambda in the signature section, but if you think it would benefit from increased readability I can have a go at trying to rewrite it more verbosely?

topliceanu commented:
@hot-leaf-juice No, there's no strong reason to do it. It's good to go!
